sdk/core/System.ClientModel/src/Pipeline/ClientPipeline.RequestOptionsProcessor.cs

```diff
@@ -0,0 +1,225 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace System.ClientModel.Primitives;
+
+public partial class ClientPipeline
+{
+    /// <summary>
+    /// Pipeline processor to advance through policies for pipeline customized
+    /// per-request by passing RequestOptions to a protocol method.
+    /// </summary>
+    internal class RequestOptionsProcessor : IReadOnlyList<PipelinePolicy>
+    {
+        private readonly int _perCallIndex;
+        private readonly int _perTryIndex;
+        private readonly int _beforeTransportIndex;
+        private readonly int _length;
+
+        // Original client-scope pipeline.
+        private readonly ReadOnlyMemory<PipelinePolicy> _fixedPolicies;
+
+        // Custom per-call policies used for the scope of the method invocation.
+        private readonly ReadOnlyMemory<PipelinePolicy> _customPerCallPolicies;
+
+        // Custom per-try policies used for the scope of the method invocation.
+        private readonly ReadOnlyMemory<PipelinePolicy> _customPerTryPolicies;
+
+        // Custom per-try policies used for the scope of the method invocation.
+        private readonly ReadOnlyMemory<PipelinePolicy> _customBeforeTransportPolicies;
+
+        private PolicyEnumerator? _enumerator;
+
+        public RequestOptionsProcessor(
+            ReadOnlyMemory<PipelinePolicy> fixedPolicies,
+            ReadOnlyMemory<PipelinePolicy> perCallPolicies,
+            ReadOnlyMemory<PipelinePolicy> perTryPolicies,
+            ReadOnlyMemory<PipelinePolicy> beforeTransportPolicies,
+            int perCallIndex,
+            int perTryIndex,
+            int beforeTransportIndex)
```

Comment:
Curious - could these be derived from the supplied policies? As I understand it, they would always be the boundary indexes between the respective policy collections.

Reply:
This is probably possible.  Since we create this in the pipeline.Send method, though, I don't think we want to move move computation into that method -- pipeline.Send should be optimized for perf, whereas pipeline construction can take a bit longer.  If you think the current design has some disadvantages, though, let me know and we can see if there are other ways to address them!

Reply:
My main concern is that at best we are asking for information we already have, and at worst we allow for these indexes to be wrong.

Reply:
Regarding perf, the validation only happens in debug builds.  I think we need to be mindful of perf here, at the possible cost of making a mistake (which at least can provide test coverage to address?).

If you have a proposal for how we could get those indexes in a high-performance way, I'm happy to consider it!